### ISSUE TITLE
`conan install` corrupts the cache and prevents other projects from compiling

### ISSUE THREAD

**Issue by Blinue** [conan-io/conan#14933]
### Environment details

* Operating System+version: Windows 11 22H2
* Compiler+version: Visual Studio 2022
* Conan version: 2.0.13
* Python version: 3.11.2

### Steps to reproduce

We are migrating from Conan v1 to v2 and encountered a bug. Here is a simple demonstration: [ConanCache.zip](https://github.com/conan-io/conan/files/12904230/ConanCache.zip).

conanfile.txt:
```
[requires]
fmt/10.1.1

[generators]
MSBuildDeps

[options]
fmt/*:header_only=True
```
First, we compile the dependencies for Project1, and Project1 works fine. But after we compile the dependencies for Project2, Project1 fails to compile because it cannot find <fmt/format.h>.

This error only occurs under specific configurations. If we remove the `header_only` option or use `--build=missing` in `conan install`, everything works fine.

**Comment by memsharded** [comment]
Hi @Blinue 

Thanks for your report and thanks for providing code to try to reproduce it.
I have been trying to reproduce, but couldn't make it fail so far.
I had to to some minor changes, like adding ``MSBuildToolchain`` to the generators, and then calling ``Project1/.conan/conanbuild.bat`` in order to activate the vcvars, otherwise the ``msbuild ...`` command fails with ``msbuild not found`` as expected, because msbuild is not in the path by default.

But once I did that, it seems it built successfully.

As a side note, I wouldn't recommend to use ``--build=*``, because it unnecessarily re-build dependencies from source again and again, taking time and storage. The recommended would be ``--build=missing`` for testing and experimenting and not using ``--build`` in production (assuming that the binaries have already been built and available in the servers, or failing otherwise).

Not sure what could be failing. I also have VS2022. How are you running ``msbuild``? Maybe you can paste your whole output of running ``build.bat``, to see if there is something there that could give us some hint.

**Comment by Blinue** [comment]
Hello, thank you for your response! Please run build.bat in the "Visual Studio 2022 Developer Command Prompt", or you can temporarily add msbuild to the path.

> As a side note, I wouldn't recommend to use --build=*, because it unnecessarily re-build dependencies from source again and again, taking time and storage.

We are using Directory.Build.props to customize the behavior of msbuild, so we have to compile all the dependencies ourselves.

Here is the output:

```
C:\Users\Xu\Downloads\ConanCache>conan install conanfile.txt --output-folder Project1\.conan --build=* -s build_type=Release -s arch=x86_64 --update

======== Input profiles ========
Profile host:
[settings]
arch=x86_64
build_type=Release
compiler=msvc
compiler.cppstd=14
compiler.runtime=dynamic
compiler.runtime_type=Release
compiler.version=193
os=Windows

Profile build:
[settings]
arch=x86_64
build_type=Release
compiler=msvc
compiler.cppstd=14
compiler.runtime=dynamic
compiler.runtime_type=Release
compiler.version=193
os=Windows


======== Computing dependency graph ========
fmt/10.1.1: Checking remote: conancenter
Graph root
    conanfile.txt: C:\Users\Xu\Downloads\ConanCache\conanfile.txt
Requirements
    fmt/10.1.1#cd63809a79574a2f9eb73ca35f16a243 - Cache (conancenter)

======== Computing necessary packages ========
fmt/10.1.1: Forced build from source
Requirements
    fmt/10.1.1#cd63809a79574a2f9eb73ca35f16a243:da39a3ee5e6b4b0d3255bfef95601890afd80709 - Build

======== Installing packages ========

-------- Installing package fmt/10.1.1 (1 of 1) --------
fmt/10.1.1: Building from source
fmt/10.1.1: Package fmt/10.1.1:da39a3ee5e6b4b0d3255bfef95601890afd80709
fmt/10.1.1: Copying sources to build folder
fmt/10.1.1: Building your package in C:\Users\Xu\.conan2\p\b\fmtacd454cb5451f\b
fmt/10.1.1: Calling generate()
fmt/10.1.1: Generators folder: C:\Users\Xu\.conan2\p\b\fmtacd454cb5451f\b\build-release\conan
fmt/10.1.1: Generating aggregated env files
fmt/10.1.1: Generated aggregated env files: ['conanbuild.bat', 'conanrun.bat']
fmt/10.1.1: Calling build()
fmt/10.1.1: apply_conandata_patches(): No patches defined in conandata
fmt/10.1.1: Package 'da39a3ee5e6b4b0d3255bfef95601890afd80709' built
fmt/10.1.1: Build folder C:\Users\Xu\.conan2\p\b\fmtacd454cb5451f\b\build-release
fmt/10.1.1: Generating the package
fmt/10.1.1: Packaging in folder C:\Users\Xu\.conan2\p\b\fmtacd454cb5451f\p
fmt/10.1.1: Calling package()
fmt/10.1.1: package(): Packaged 13 '.h' files
fmt/10.1.1: package(): Packaged 1 '.rst' file: LICENSE.rst
fmt/10.1.1: Created package revision 6cd13a7517e9261022d9b2b853057129
fmt/10.1.1: Package 'da39a3ee5e6b4b0d3255bfef95601890afd80709' created
fmt/10.1.1: Full package reference: fmt/10.1.1#cd63809a79574a2f9eb73ca35f16a243:da39a3ee5e6b4b0d3255bfef95601890afd80709#6cd13a7517e9261022d9b2b853057129
fmt/10.1.1: Package folder C:\Users\Xu\.conan2\p\b\fmtacd454cb5451f\p
WARN: deprecated: Usage of deprecated Conan 1.X features that will be removed in Conan 2.X:
WARN: deprecated:     'cpp_info.names' used in: fmt/10.1.1

======== Finalizing install (deploy, generators) ========
conanfile.txt: Writing generators to C:\Users\Xu\Downloads\ConanCache\Project1\.conan
conanfile.txt: Generator 'MSBuildDeps' calling 'generate()'
conanfile.txt: Generating aggregated env files
conanfile.txt: Generated aggregated env files: ['conanbuild.bat', 'conanrun.bat']
Install finished successfully

C:\Users\Xu\Downloads\ConanCache>REM WORKS

C:\Users\Xu\Downloads\ConanCache>msbuild -p:Configuration=Release;Platform=x64 Project1\Project1.sln
MSBuild version 17.7.2+d6990bcfa for .NET Framework
Build started 2023/10/15 9:20:08.

Project "C:\Users\Xu\Downloads\ConanCache\Project1\Project1.sln" on node 1 (default targets).
ValidateSolutionConfiguration:
  Building solution configuration "Release|x64".
Project "C:\Users\Xu\Downloads\ConanCache\Project1\Project1.sln" (1) is building "C:\Users\Xu\Downloads\ConanCache\Proj
ect1\Project1.vcxproj" (2) on node 1 (default targets).
InitializeBuildStatus:
  Touching "x64\Release\Project1.tlog\unsuccessfulbuild".
ClCompile:
  C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.37.32822\bin\HostX64\x64\CL.exe /c /IC:/User
  s/Xu/.conan2/p/b/fmtacd454cb5451f/p/include /Zi /nologo /W3 /WX- /diagnostics:column /sdl /O2 /Oi /GL /D FMT_HEADER_O
  NLY=1 /D NDEBUG /D _CONSOLE /D _UNICODE /D UNICODE /Gm- /EHsc /MD /GS /Gy /fp:precise /Zc:wchar_t /Zc:forScope /Zc:in
  line /permissive- /Fo"x64\Release\\" /Fd"x64\Release\vc143.pdb" /external:W3 /Gd /TP /FC /errorReport:queue  Project1
  .cpp
  Project1.cpp
Link:
  C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.37.32822\bin\HostX64\x64\link.exe /ERRORREPO
  RT:QUEUE /OUT:"C:\Users\Xu\Downloads\ConanCache\Project1\x64\Release\Project1.exe" /NOLOGO kernel32.lib user32.lib gd
  i32.lib winspool.lib comdlg32.lib advapi32.lib shell32.lib ole32.lib oleaut32.lib uuid.lib odbc32.lib odbccp32.lib /M
  ANIFEST /MANIFESTUAC:"level='asInvoker' uiAccess='false'" /manifest:embed /DEBUG /PDB:"C:\Users\Xu\Downloads\ConanCac
  he\Project1\x64\Release\Project1.pdb" /SUBSYSTEM:CONSOLE /OPT:REF /OPT:ICF /LTCG:incremental /LTCGOUT:"x64\Release\Pr
  oject1.iobj" /TLBID:1 /DYNAMICBASE /NXCOMPAT /IMPLIB:"C:\Users\Xu\Downloads\ConanCache\Project1\x64\Release\Project1.
  lib" /MACHINE:X64  x64\Release\Project1.obj
  ??????
  853 of 1121 functions (76.1%) were compiled, the rest were copied from previous compilation.
    0 functions were new in current compilation
    0 functions had inline decision re-evaluated but remain unchanged
  ????????
  Project1.vcxproj -> C:\Users\Xu\Downloads\ConanCache\Project1\x64\Release\Project1.exe
FinalizeBuildStatus:
  Deleting file "x64\Release\Project1.tlog\unsuccessfulbuild".
  Touching "x64\Release\Project1.tlog\Project1.lastbuildstate".
Done Building Project "C:\Users\Xu\Downloads\ConanCache\Project1\Project1.vcxproj" (default targets).

Done Building Project "C:\Users\Xu\Downloads\ConanCache\Project1\Project1.sln" (default targets).


Build succeeded.
    0 Warning(s)
    0 Error(s)

Time Elapsed 00:00:01.79

C:\Users\Xu\Downloads\ConanCache>conan install conanfile.txt --output-folder Project2\.conan --build=* -s build_type=Release -s arch=x86_64 --update

======== Input profiles ========
Profile host:
[settings]
arch=x86_64
build_type=Release
compiler=msvc
compiler.cppstd=14
compiler.runtime=dynamic
compiler.runtime_type=Release
compiler.version=193
os=Windows

Profile build:
[settings]
arch=x86_64
build_type=Release
compiler=msvc
compiler.cppstd=14
compiler.runtime=dynamic
compiler.runtime_type=Release
compiler.version=193
os=Windows


======== Computing dependency graph ========
fmt/10.1.1: Checking remote: conancenter
Graph root
    conanfile.txt: C:\Users\Xu\Downloads\ConanCache\conanfile.txt
Requirements
    fmt/10.1.1#cd63809a79574a2f9eb73ca35f16a243 - Cache (conancenter)

======== Computing necessary packages ========
fmt/10.1.1: Forced build from source
Requirements
    fmt/10.1.1#cd63809a79574a2f9eb73ca35f16a243:da39a3ee5e6b4b0d3255bfef95601890afd80709 - Build

======== Installing packages ========

-------- Installing package fmt/10.1.1 (1 of 1) --------
fmt/10.1.1: Building from source
fmt/10.1.1: Package fmt/10.1.1:da39a3ee5e6b4b0d3255bfef95601890afd80709
fmt/10.1.1: Copying sources to build folder
fmt/10.1.1: Building your package in C:\Users\Xu\.conan2\p\b\fmta03388bb88e4b\b
fmt/10.1.1: Calling generate()
fmt/10.1.1: Generators folder: C:\Users\Xu\.conan2\p\b\fmta03388bb88e4b\b\build-release\conan
fmt/10.1.1: Generating aggregated env files
fmt/10.1.1: Generated aggregated env files: ['conanbuild.bat', 'conanrun.bat']
fmt/10.1.1: Calling build()
fmt/10.1.1: apply_conandata_patches(): No patches defined in conandata
fmt/10.1.1: Package 'da39a3ee5e6b4b0d3255bfef95601890afd80709' built
fmt/10.1.1: Build folder C:\Users\Xu\.conan2\p\b\fmta03388bb88e4b\b\build-release
fmt/10.1.1: Generating the package
fmt/10.1.1: Packaging in folder C:\Users\Xu\.conan2\p\b\fmta03388bb88e4b\p
fmt/10.1.1: Calling package()
fmt/10.1.1: package(): Packaged 13 '.h' files
fmt/10.1.1: package(): Packaged 1 '.rst' file: LICENSE.rst
fmt/10.1.1: Created package revision 6cd13a7517e9261022d9b2b853057129
fmt/10.1.1: Package 'da39a3ee5e6b4b0d3255bfef95601890afd80709' created
fmt/10.1.1: Full package reference: fmt/10.1.1#cd63809a79574a2f9eb73ca35f16a243:da39a3ee5e6b4b0d3255bfef95601890afd80709#6cd13a7517e9261022d9b2b853057129
fmt/10.1.1: Package folder C:\Users\Xu\.conan2\p\b\fmta03388bb88e4b\p
WARN: deprecated: Usage of deprecated Conan 1.X features that will be removed in Conan 2.X:
WARN: deprecated:     'cpp_info.names' used in: fmt/10.1.1

======== Finalizing install (deploy, generators) ========
conanfile.txt: Writing generators to C:\Users\Xu\Downloads\ConanCache\Project2\.conan
conanfile.txt: Generator 'MSBuildDeps' calling 'generate()'
conanfile.txt: Generating aggregated env files
conanfile.txt: Generated aggregated env files: ['conanbuild.bat', 'conanrun.bat']
Install finished successfully

C:\Users\Xu\Downloads\ConanCache>REM FAIL!!!

C:\Users\Xu\Downloads\ConanCache>msbuild -p:Configuration=Release;Platform=x64 Project1\Project1.sln
MSBuild version 17.7.2+d6990bcfa for .NET Framework
Build started 2023/10/15 9:20:11.

Project "C:\Users\Xu\Downloads\ConanCache\Project1\Project1.sln" on node 1 (default targets).
ValidateSolutionConfiguration:
  Building solution configuration "Release|x64".
Project "C:\Users\Xu\Downloads\ConanCache\Project1\Project1.sln" (1) is building "C:\Users\Xu\Downloads\ConanCache\Proj
ect1\Project1.vcxproj" (2) on node 1 (default targets).
InitializeBuildStatus:
  Creating "x64\Release\Project1.tlog\unsuccessfulbuild" because "AlwaysCreate" was specified.
  Touching "x64\Release\Project1.tlog\unsuccessfulbuild".
ClCompile:
  C:\Program Files\Microsoft Visual Studio\2022\Community\VC\Tools\MSVC\14.37.32822\bin\HostX64\x64\CL.exe /c /IC:/User
  s/Xu/.conan2/p/b/fmtacd454cb5451f/p/include /Zi /nologo /W3 /WX- /diagnostics:column /sdl /O2 /Oi /GL /D FMT_HEADER_O
  NLY=1 /D NDEBUG /D _CONSOLE /D _UNICODE /D UNICODE /Gm- /EHsc /MD /GS /Gy /fp:precise /Zc:wchar_t /Zc:forScope /Zc:in
  line /permissive- /Fo"x64\Release\\" /Fd"x64\Release\vc143.pdb" /external:W3 /Gd /TP /FC /errorReport:queue  Project1
  .cpp
  Project1.cpp
C:\Users\Xu\Downloads\ConanCache\Project1\Project1.cpp(2,10): fatal  error C1083: ????????: "fmt/format.h": No such fil
e or directory [C:\Users\Xu\Downloads\ConanCache\Project1\Project1.vcxproj]
Done Building Project "C:\Users\Xu\Downloads\ConanCache\Project1\Project1.vcxproj" (default targets) -- FAILED.

Done Building Project "C:\Users\Xu\Downloads\ConanCache\Project1\Project1.sln" (default targets) -- FAILED.


Build FAILED.

"C:\Users\Xu\Downloads\ConanCache\Project1\Project1.sln" (default target) (1) ->
"C:\Users\Xu\Downloads\ConanCache\Project1\Project1.vcxproj" (default target) (2) ->
(ClCompile target) ->
  C:\Users\Xu\Downloads\ConanCache\Project1\Project1.cpp(2,10): fatal  error C1083: ????????: "fmt/format.h": No such f
ile or directory [C:\Users\Xu\Downloads\ConanCache\Project1\Project1.vcxproj]

    0 Warning(s)
    1 Error(s)

Time Elapsed 00:00:00.51
```

**Comment by Blinue** [comment]
When running `conan install` for the second time, `~/.conan2/p/b/fmtacd454cb5451f` was deleted, causing Project1 to fail to compile.

**Comment by memsharded** [comment]
Thanks for the details. Running the ``build.bat`` from inside a VS2022 prompt still works fine in my computer,  

But I think I have some new hints to consider:
- The ``--build=*`` is actually rebuilding from source exactly the same thing, because it is a header-only, so it computes again the same package-revision and it updates the DB to point to the new folder location
- I have reviewed and debugged relevant parts of the code, and I cannot find anywhere were the old package-revision folder is being removed
- Are you sure you reproduce with exactly the same ``build.bat`` script? Aren't you using any other intermediate command, like a ``conan cache clean``, ``conan remove`` or something like that? Any other possible operation on the Conan cache?


> We are using Directory.Build.props to customize the behavior of msbuild, so we have to compile all the dependencies ourselves.

``fmt`` (and the majority of packages from ConanCenter) doesn't use MSBuild to build, so I'd say it is irrelevant what ``Directory.Build.props`` you use, or am I missing something? The way the customization of dependencies happen is via toolchains, which is defined by Conan ``settings`` and ``options``.

In any case, it is still unnecessary and counter-productive to call ``--build=*`` all the time, it is unnecessarily rebuilding from source the same packages again and again. In this case it is only fmt header-only which is fast and not noticeable, but the moment you add other dependencies, it will have a much larger impact. You might want to build your dependencies once first (for example ``conan install --requires=fmt/... --build=fmt/*``, upload them to your own server and use them from there without ``--build``. Please check the guide https://docs.conan.io/2/devops/using_conancenter.html

Please let me know about the above ``build.bat`` questions, there might be some little detail that we are missing there that could be relevant.

**Comment by memsharded** [comment]
I keep investigating in the codebase, running some local tests, checking the cache DB, etc, I will keep you posted about any further finding.

**Comment by memsharded** [comment]
Ok, I have managed to reproduce it. For some reason my sqlite3 installation was not behaving correctly, re-installing things and I now can see what is happening. Thanks again, we will be checking how this can be improved.

**Comment by Blinue** [comment]
Thank you for your investigation. I just created a [repository](https://github.com/Blinue/Conan-Issue) to demonstrate this error, but you probably don’t need it anymore :)

**Comment by memsharded** [comment]
Thanks for that repo too!

It took me a while to fully grasp what was happening and finding an approach that hopefully will not break other usages and users. I am proposing the fix in https://github.com/conan-io/conan/pull/14938

In any case, please note that the trigger of this was the unexpected ``--build=*``, forcing rebuilding from source again and again the same package with exact same package revision (header-only). Please try to avoid rebuilding from source in production, the only recommended approaches would be ``--build=missing``, or depending the use case, maybe just ``conan create`` should be enough to build binaries, and all other ``conan install`` should re-use those binaries.

**Comment by memsharded** [comment]
Hi @Blinue 

https://github.com/conan-io/conan/pull/14938 has been merged, and it will be in next 2.0.14. Thanks again for your feedback!
